### DISCUSSION TITLE
V2: Levels & Achievements

### DISCUSSION THREAD

**Discussion by 0x4007** [ubiquity/ubiquibot#71] [Specifications]
We want to gamify and recognize talent as they continue to climb the ranks with hunting bounties. Some examples could be:
- Bounty Hunter could level up to Bounty Master if they close one bounty per week for minimum four weeks.
- We could give an achievement "communicator" if the user responds within 24 hours to every comment in conversations on PRs they are apart of. 
- We could give an achievement "timely" if the Bounty Hunter has closed two bounties on time.

These are just off the cuff and certainly will be refined over time. It all comes down to our ability to collect data on user behavior and assign levels/achievements once certain conditions are met. 

---

### Data to collect

1. Number of bounties completed (XP)
2. Number of bounties started (Success Rate) 
3. Dollars collected (Earned)
4. Number of Repositories contributed to (Repos)
5. Types of bounties completed e.g. UI/UX; DevOps; Smart Contracts; Backend (Type of Dev)
6. Commit timestamps 
7. Bounty completion time percentage difference from Bounty Expiration time (Timeliness)

### Useful to Calculate and Display

- Average bounties solved per month
- Average dollars collected per month
- Average daily commits
- Average timeliness (i.e. negative percentage means on average finishes within the allotted bounty time)

**Comment by 0x4007** [comment]
I think after a bounty hunter has an XP of greater than zero, meaning they successfully completed a bounty, they should be eligible for financial incentives like getting paid to review PRs and leaving comments. Requires a bit more research to see how this can be exploited but seems to make sense to exclude randoms.

**Comment by Draeieg** [comment]
Bounty hunters can gain xp and level up, allowing them to complete higher priority bounties 

xp=dollars earned
if a bounty is performed perfectly the hunter will get a perfect "bonus" on his xp (maybe to the dollars too), this bonus can turn into a "streak" where each successive perfect bounty will grow the streak and as a result the bonus can grow too

this means hunters that complete many small tasks perfectly will get more rewards than hunters that only complete bigger bounties, while hunters can do that  this will numerically show the difference between those who read and work carefully, effectively "Gaming the system"

**Reply by 0x4007** [reply]
I'm not sure if this is the best place to post this but you made me think of something:

What if a reviewer's financial incentive is independently calculated and tallied up. If the compensation to the reviewing team is higher than the original bounty, the bot comments a warning message stating that the reviewer compensation is maxed out for that pull request, and that it's likely to be deprioritized (which could pressure the bounty hunter to produce higher quality pull requests?) 

It is a simplification from our earlier "tennis match" concept.

**Comment by 0x4007** [comment]
How to algorithmically derive reputation of projects: https://github.com/ubiquity/ubiquibot/issues/441#issuecomment-1611752748
